Add Fake TNT and Random Teleport perks

Luncaaa opened the pull request:
(From Dol's suggestions at the end of the main file)

Fake TNT is the same as the normal TNT perk but it won't explode at the end and the random teleport perk will teleport you to any random and solid block inside the arena.

The cuboid area has a new property which is the list of "safe blocks" in the arena. These are existing solid blocks where the player can be teleported. When a block is broken by a player, TNT or arena shrinking, the block is removed from the list.

Also, the plugin won't throw an error if the "DegradeOnTick" is 0.

tierrif commented:
```
[18:32:28 ERROR]: Could not pass event PlayerInteractEvent to FloorIsLava v2.9.1
java.lang.IllegalArgumentException: bound must be positive
        at java.util.Random.nextInt(Random.java:322) ~[?:?]
        at com.gmail.tracebachi.floorislava.utils.CuboidArea.getRandomSafeLocationInside(CuboidArea.java:120) ~[FloorIsLava-2.9.1.jar:?]
        at com.gmail.tracebachi.floorislava.arena.perks.RandomTp.onPerkActivation(RandomTp.java:22) ~[FloorIsLava-2.9.1.jar:?]
        at com.gmail.tracebachi.floorislava.arena.perks.Perk.activate(Perk.java:42) ~[FloorIsLava-2.9.1.jar:?]
        at com.gmail.tracebachi.floorislava.arena.Arena.onPlayerBlockClick(Arena.java:461) ~[FloorIsLava-2.9.1.jar:?]
        at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor113.execute(Unknown Source) ~[?:?]
        at org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:75) ~[paper-api-1.18.2-R0.1-SNAPSHOT.jar:?]
        at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80) ~[paper-api-1.18.2-R0.1-SNAPSHOT.jar:git-Paper-388]
        at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[paper-api-1.18.2-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:669) ~[paper-api-1.18.2-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPlayerInteractEvent(CraftEventFactory.java:544) ~[paper-1.18.2.jar:git-Paper-388]
        at net.minecraft.server.level.ServerPlayerGameMode.useItemOn(ServerPlayerGameMode.java:563) ~[?:?]
        at net.minecraft.server.network.ServerGamePacketListenerImpl.handleUseItemOn(ServerGamePacketListenerImpl.java:1852) ~[?:?]
        at net.minecraft.network.protocol.game.ServerboundUseItemOnPacket.handle(ServerboundUseItemOnPacket.java:33) ~[?:?]
        at net.minecraft.network.protocol.game.ServerboundUseItemOnPacket.a(ServerboundUseItemOnPacket.java:9) ~[?:?]
        at net.minecraft.network.protocol.PacketUtils.lambda$ensureRunningOnSameThread$1(PacketUtils.java:51) ~[?:?]
        at net.minecraft.server.TickTask.run(TickTask.java:18) ~[paper-1.18.2.jar:git-Paper-388]
        at net.minecraft.util.thread.BlockableEventLoop.doRunTask(BlockableEventLoop.java:153) ~[?:?]
        at net.minecraft.util.thread.ReentrantBlockableEventLoop.doRunTask(ReentrantBlockableEventLoop.java:24) ~[?:?]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:1400) ~[paper-1.18.2.jar:git-Paper-388]
        at net.minecraft.server.MinecraftServer.d(MinecraftServer.java:188) ~[paper-1.18.2.jar:git-Paper-388]
        at net.minecraft.util.thread.BlockableEventLoop.pollTask(BlockableEventLoop.java:126) ~[?:?]
        at net.minecraft.server.MinecraftServer.pollTaskInternal(MinecraftServer.java:1377) ~[paper-1.18.2.jar:git-Paper-388]
        at net.minecraft.server.MinecraftServer.pollTask(MinecraftServer.java:1370) ~[paper-1.18.2.jar:git-Paper-388]
        at net.minecraft.util.thread.BlockableEventLoop.managedBlock(BlockableEventLoop.java:136) ~[?:?]
        at net.minecraft.server.MinecraftServer.waitUntilNextTick(MinecraftServer.java:1348) ~[paper-1.18.2.jar:git-Paper-388]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1230) ~[paper-1.18.2.jar:git-Paper-388]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:316) ~[paper-1.18.2.jar:git-Paper-388]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

I get this in an arena with a 1-block thick map, seems to be an issue with your safe block functionality. Random TP ends up not working at all. I'll assume this also happens when the arena fully decayed if it can't find no safe blocks either. But in this case, it's finding none, even when they exist.

Luncaaa commented:
The problem was that blocks broken on a game wouldn't count as safe blocks on the next game. I fixed this by creating another list where blocks aren't removed and, at the end of a game, the list is cloned into the safe blocks list.

Also, I was not able to join 1 block wide arenas because of an error not related to what I implemented, so I fixed it as well

tierrif commented:
Functionality seems to be fine. Awaiting code review (will do it myself if @AstelonM doesn't want to).
~~I have a suggestion that shouldn't be implemented yet, but I just think may be a little better: Make the fake TNT's icon an enchanted TNT, instead of a redstone lamp which doesn't make too much sense.~~ <-- Forget this, too hard to implement

When it comes to implementation in the server, we will need to tweak perks a bit in order to not have the teleporter be too OP. However, from what I tested, it doesn't seem like a perfect alternative to boost, which may mean it doesn't need nerfing. This will need more testing but is irrelevant to this PR.

AstelonM commented on src/main/java/com/gmail/tracebachi/floorislava/gui/FloorGuiMenu.java:
```diff
@@ -42,13 +41,15 @@ public class FloorGuiMenu {
     public static final ItemStack BOOST_ITEM = new ItemStack(Material.FEATHER);
     public static final ItemStack CHIKUN_ITEM = new ItemStack(Material.EGG);
     public static final ItemStack STEAL_ITEM = new ItemStack(Material.FLINT_AND_STEEL);
+    public static final ItemStack RANDOM_TP_ITEM = new ItemStack(Material.CHORUS_FRUIT);
```
Either change the material or make sure it can't be eaten

AstelonM replied:
This was confirmed to be working properly

AstelonM commented on src/main/java/com/gmail/tracebachi/floorislava/utils/CuboidArea.java:
Doesn't this also make it so the watch area has safe blocks computed as well? Shouldn't affect anything, but rather redundant

Luncaaa commented:
> ~I have a suggestion that shouldn't be implemented yet, but I just think may be a little better: Make the fake TNT's icon an enchanted TNT, instead of a redstone lamp which doesn't make too much sense.~ <-- Forget this, too hard to implement

I have done this but it could be confusing. When selecting a perk, it will have the enchanted effect to tell the player that he has it selected, but having an enchanted TNT could look as if he had selected it. The same happens when selecting the fake tnt once: you have no way of knowing if you have it selected or not because it's always glowing.
(The block would be TNT instead of a redstone lamp in the images)
![example](https://user-images.githubusercontent.com/80987436/233697767-0565c2a5-fd64-4a47-a805-ba5e5a342b09.png)
![example2](https://user-images.githubusercontent.com/80987436/233697771-28f40967-06c1-42c2-835b-995712188c8d.png)

tierrif commented:
That's a fair point too, let's just keep it as-is for now and focus on the review changes

AstelonM reviewed (approved):
Looks good now
